### PR TITLE
Prototyping v2

### PR DESCRIPTION
**Pull request by tin1254**

### CODE REVIEW

**Review by tin1254 — COMMENTED** [review]
Removed

File: voxel_grid/prototype/mvp_v2.hpp [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+  }
+
+  size_t getMinimumPointsNumberPerVoxel() const {
+    return min_points_per_voxel_;
+  }
+
+ protected:
+  Vec3f leaf_size_ = Eigen::Vector3f::Constant(0.05);
+  size_t min_points_per_voxel_;
+  std::vector<int> leaf_layout_;
```
Comment (kunaltyagi) [inline voxel_grid/prototype/mvp_v2.hpp]: Why is this required in GridFilterBase?

File: voxel_grid/prototype/mvp_v2.hpp [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    output.reserve(grid_struct_.size());
+    for (auto it = grid_struct_.begin(); it != grid_struct_.end(); ++it) {
+      const auto &res = grid_struct_.filterGrid(it);
+      if (res) {
+        output.push_back(res.value());
+      }
+    }
+
+    // push back leaf_layout_ to the base
+    grid_struct_.saveResults(this);
```
Comment (kunaltyagi) [inline voxel_grid/prototype/mvp_v2.hpp]: We can prevent the static_cast and this function requirement for all filters by moving this into `void GridFilterBase::applyFilter(PointCloud& out) override`

File: voxel_grid/prototype/mvp_v2.hpp [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+template <typename GridStruct, typename PointT = GET_POINT_TYPE(GridStruct)>
+class TransformFilter : public Filter<PointT> {
+  using PointCloud = PointCloudT<PointXYZ>;
+
+ protected:
+  TransformFilter(GridStruct grid_struct) {
+    grid_struct_ = std::move(grid_struct);
+  }
+
+  void applyFilter(PointCloud &output) {
```
Comment (kunaltyagi) [inline voxel_grid/prototype/mvp_v2.hpp]:
```suggestion
  void applyFilter(PointCloud &output) override {
```

File: voxel_grid/prototype/mvp_v2.hpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+  Vec3f inverse_leaf_size_;
+  size_t min_points_per_voxel_;
+  Grid grid_;
+
+  size_t num_centroids = 0;
+  std::vector<int> leaf_layout_;
+
+ private:
+  template <typename, typename>
+  friend class TransformFilter;
```
Comment (kunaltyagi) [inline voxel_grid/prototype/mvp_v2.hpp]: What hidden data do we want to access from transformFilter?

File: voxel_grid/prototype/mvp_v2.hpp (hunk 2 above)
Comment (tin1254) [reply voxel_grid/prototype/mvp_v2.hpp:87]: I can try it out. 

Then the purposes of `TransformFilter` are just holding `GridStruct` object and accessing the data in `GridStruct` right?

File: voxel_grid/prototype/mvp_v2.hpp (hunk 4 above)
Comment (tin1254) [reply voxel_grid/prototype/mvp_v2.hpp]: We don't need it actually, but for hiding some functions from public. I'll remove it

File: voxel_grid/prototype/mvp_v2.hpp (hunk 3 above)
Comment (tin1254) [reply voxel_grid/prototype/mvp_v2.hpp]: Done

File: voxel_grid/prototype/mvp_v2.hpp (hunk 1 above)
Comment (tin1254) [reply voxel_grid/prototype/mvp_v2.hpp]: Removed

File: voxel_grid/prototype/mvp_v2.hpp (hunk 2 above)
Comment (tin1254) [reply voxel_grid/prototype/mvp_v2.hpp]: I removed `saveResults` so I'll keep `applyFilter` in `TransformFilter` for now (`static_cast` is still needed)

File: voxel_grid/prototype/mvp_v2.hpp [hunk 5]
```diff
@@ -82,9 +82,15 @@ class TransformFilter : public Filter<PointT> {
...
+
+  // need to declare the corresponding functions in GridStruct
+  std::vector<int> getLeafLayout() const {
+    return grid_struct_.getLeafLayout();
+  }
 
-    // push back leaf_layout_ to the base
-    grid_struct_.saveResults(this);
+  inline int getCentroidIndex(float x, float y, float z) const {
+    return grid_struct_.getCentroidIndex(x, y, z);
```
Comment (kunaltyagi) [inline voxel_grid/prototype/mvp_v2.hpp]: Can we move these to GridFilterBase and instead replace them with `getTransformer` or similar to return reference to `grid_struct_`?

File: voxel_grid/prototype/mvp_v2.hpp [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+  virtual void applyFilter(PointCloud &pc) = 0;
+  PointCloud input_;
+};
+
+// ================== transform_filter.hpp ==================
+
+#define GET_POINT_TYPE(GridStructT) typename GridStructT::PointCloud::PointType
+
+template <typename GridStruct, typename PointT = GET_POINT_TYPE(GridStruct)>
+class TransformFilter : public Filter<PointT> {
```
Comment (kunaltyagi) [inline voxel_grid/prototype/mvp_v2.hpp:62]: As mentioned by me elsewhere, we should create a uniform interface for template types for FunctionFilter and TransformFilter (both should have either PointT  first or second). Smthg to discuss, but not here (just mentioning so I don't forget it)

File: voxel_grid/prototype/mvp_v2.hpp (hunk 5 above)
Comment (tin1254) [reply voxel_grid/prototype/mvp_v2.hpp]: Yes, we can have a function returning the reference.

But we have a few functions which are specific to certain filters, like `getCentroidIndex` is only for `VoxelGrid` and `VoxelGridLabel`. (we have 5 grid filters in total) So adding those functions to either `GridFilterBase` or `TransformFilter` I think is implausible.
If we want to keep APIs of `VoxelGrid` the same as before, the only way I can think of is to add another class to declare those filter specific functions. 

@kunaltyagi Do you have any better idea to get around that?

Comment (kunaltyagi) [reply voxel_grid/prototype/mvp_v2.hpp]: Those functions can go in the last tier (transform->gridBase->VoxelFilter) right?

Comment (tin1254) [reply voxel_grid/prototype/mvp_v2.hpp]: Done. I created a new class as `VoxelFilter` and `VoxelGrid` is once again declared by `using`

## Brief recap

### Goal of this prototype

> Pro
>     - Prevents multiple inheritance
>     - Restricts GridStruct from adding any API/modifying functionality
>
>  Cons
>     - Requires used to sub-class to add functionality

Seems like we achieved both of the pro

### Initial goal for grid filters

  [✓] Reduce boilerplate code
  [✓] Allow users pass their `GridStruct` to have custom grid filter
  [✓] Allow upgrading concurrence with ease

### Comments
1. Reduce boilerplate code
    - [✓]  boilerplate logic in `applyFilter` is removed
    - [✓] Not all common code is reduced
        - Setting up `min_b_` and `max_b_` are common, but they are just a few lines of code (usually >10 lines)
        - Hashing is basically universal
            - **Done**: move to `GridFilterBase`
 2. Allow users pass their `GridStruct` to have custom grid filter
    - [✓] Users can pass their `GridStruct` to `TransformFilter`
 3. Allow upgrading concurrence with ease
     - [✓] All for loops are located in `applyFilter` now, all grid filters including filters custom `GridStruct` can be benefit from concurrence

**Comment by tin1254** [comment]
Design is adopted

**Comment by tin1254** [comment]
Design is adopted
